### ISSUE TITLE
Nicer mode change messages

### ISSUE THREAD

**Issue by xPaw** [thelounge/lounge#84]

**Comment by astorije** [comment]
:+1:
What about `xPaw has granted test channel operator (+o)`? I.e. adding the raw flag at the end in parentheses?

**Comment by dgw** [comment]
Should The Lounge really try to interpret all modes? What about nonstandard modes or modes with conflicting definitions depending on IRCd? Would The Lounge maintain a means of matching modes with the current IRCd in use?

Just to illustrate that some modes used for access have conflicting definitions, there's a pretty comprehensive list at https://www.alien.net.au/irc/chanmodes.html

**Comment by astorije** [comment]
What about only those following an RFC? Still, there are conflicting ones in the list that are coming from RFCs...

**Comment by YaManicKill** [comment]
Yeah, there's no way to know whether they are following the RFC. I like the idea, but yeah we don't want to show the wrong word, as it can be misleading.

**Comment by xPaw** [comment]
We don't have to expand every single mode into a string, just some of the most used ones, like +o, +v, +b and +q. If you look at @dgw's list, there are no conflicting implementations for these modes.

Here's how HexChat handles it: https://github.com/hexchat/hexchat/blob/78486b49d5b688ba5054631b849f8959846cf07b/src/common/modes.c#L392-L573
(they have grouped messages too)

**Comment by dgw** [comment]
+q does conflict, between the RFC definition (quiet channel) and the Unreal IRCd definition (channel owner privileges). +h doesn't conflict, though.

So would the list of modes to make "user friendly" be +v, +h, +o, and +v?

Might be worth handling ban exceptions, since the network that had a custom definition of +e considers it obsolete. +m doesn't conflict and could be expanded as well to let users know why they might not be able to talk in a channel that is set as moderated. Or is this sort of thing intended only for modes that act on users in a channel?

**Comment by xPaw** [comment]
> Or is this sort of thing intended only for modes that act on users in a channel?

Handling both would be super nice.

Looking at HexChat's code, they handle +q in both cases by checking `chanmodes` supplied by the server.

**Comment by astorije** [comment]
@xPaw, how easy/feasible is this? Also, how likely that one of us is going to look at it within a few months? Is it something we should mark as up-for-grabs?

**Comment by xPaw** [comment]
It's doable, definitely.

**Comment by McInkay** [comment]
> +q does conflict, between the RFC definition (quiet channel) and the Unreal IRCd definition (channel owner privileges). +h doesn't conflict, though.

Just noticed this.

That's not a conflict, as 1 is a channel mode and 1 is a user mode.

**Comment by RockyTV** [comment]
Should the server change the message for each mode or should the client handle it? I was thinking about checking, in the client, what mode and if it's a channel or user mode being set, then write a different message for each mode. But I'm not sure how to do it with Handlebars templates.
